op-program: Store created blocks to avoid needing to re-execute them

ajsutton opened the pull request:
**Description**

Since op-program always creates a new block via engine_forkChoiceUpdated/engine_getPayload and then immediately sends the same block back to engine_newPayload to be imported, store the created block to the database without updating the chain head, so the engine_newPayload method is a no-op instead of having to execute the transactions again.

The engine is also used in e2e tests so the block caching is optional depending on the provided backend.

ajsutton committed:
op-program: Store created blocks to avoid needing to re-execute them.

Since op-program always creates a new block via engine_forkChoiceUpdated/engine_getPayload and then immediately sends the same block back to engine_newPayload to be imported, store the created block to the database without updating the chain head, so the engine_newPayload method is a no-op instead of having to execute the transactions again.

Inphi reviewed (approved):
Basically, this ensures that the program enters [this](https://github.com/ethereum-optimism/optimism/blob/ea6941789aa0d17942d4f9e65058437be7765848/op-program/client/l2/engineapi/l2_engine_api.go#L489) if-stmt, right?

Is there a test that already asserts the cache is hit in that case?

ajsutton commented:
Test is a good callout.  I had originally enforced in in the compiler but then looked at it and wondered why I had those unnecessary changes and undid them so seems like having a test is going to be a better idea. 😆

ajsutton commented:
And yes, that's the if statement we're targeting.

ajsutton commented:
@Inphi that test got much more interesting that I first expected.  Have put up a stacked PR to make it easier to review: https://github.com/ethereum-optimism/optimism/pull/12402
I hit some surprising issues with the block hash for the created block not matching what was expected and being rejected. Those actually boiled down to the chain config used in the test being for a non-optimism chain but I've made things more consistent between the geth miner and op-program's block_processor so they do match up and we're less likely to get unexpected surprises in the future.

ajsutton committed:
op-program: Add test that created block is not reprocessed on import (#12402)

Also fixes differences in block hash for created payloads when using non-op stack chains.
